libyul/optimiser/StackCompressor.cpp

```diff
@@ -248,16 +248,22 @@ std::tuple<bool, Block> StackCompressor::run(
 		"Need to run the function grouper before the stack compressor."
 	);
 	bool usesOptimizedCodeGenerator = false;
+	std::optional<uint8_t> eofVersion = std::nullopt;
 	if (auto evmDialect = dynamic_cast<EVMDialect const*>(&_dialect))
+	{
 		usesOptimizedCodeGenerator =
 			_optimizeStackAllocation &&
 			evmDialect->evmVersion().canOverchargeGasForCall() &&
 			evmDialect->providesObjectAccess();
+
+		eofVersion = evmDialect->eofVersion();
+	}
 	bool allowMSizeOptimization = !MSizeFinder::containsMSize(_dialect, _object.code()->root());
 	Block astRoot = std::get<Block>(ASTCopier{}(_object.code()->root()));
 	if (usesOptimizedCodeGenerator)
 	{
-		yul::AsmAnalysisInfo analysisInfo = yul::AsmAnalyzer::analyzeStrictAssertCorrect(_dialect, astRoot, _object.qualifiedDataNames());
+		yul::AsmAnalysisInfo analysisInfo = yul::AsmAnalyzer::analyzeStrictAssertCorrect(_dialect, astRoot, _object.name,
+			_object.qualifiedDataNames(eofVersion));
```

Comment:
Like I said in https://github.com/ethereum/solidity/pull/15536#discussion_r1813328153, I think it would be better if `AsmAnalyzer` always received the names the way it does now (i.e. qualified) and only extracted the unqualified ones internally. Otherwise it will be just too error prone. We'll always have to remember that this parameter has a different meaning with and without EOF and I think it will be hard to keep it straight.

Also, the distinction between qualified and qualified names is already very confusing as it is. It isn't even always be reflected in naming. Making `m_dataNames` be one or the other depending on context will bring the confusion to another level. Let's not do this :)

Finally, I'm not even sure we won't need qualified names for something in EOF. We definitely don't need them for `eofcreate()` and `returncontract()`, but I don't yet know the whole picture. Maybe something will still need them.

Reply:
https://github.com/ethereum/solidity/pull/15536#discussion_r1815192103